chore: Complete README (#33)
decentraland/rentals-contract: README.md

# Rentals Contract

This contract provides on-chain renting capabilities for Decentraland's LAND and Estates.

By capitalizing on off-chain signatures, users only need to spend gas on critical transactions and avoid operations that can be done safely off-chain like handling listings or offers.

> This contract is intended for Decentraland's LAND and Estates but can be used with any ERC721 that has a `setUpdateOperator` function to give operator permissions to an address. For both of these assets, the update operator is the one that can deploy scenes on them.

Reviewer: ```suggestion
This contract is intended for Decentraland's LAND and Estates but can be used with any ERC721 that has a `setUpdateOperator` function to give operator permissions to an address. For both of these assets, the update operator can deploy scenes on them.
```

2 - Sign the rental conditions

There are various conditions that you can set for a rental to occur. You can have a glimpse of what these conditions are by looking at the Listing struct.

<a id="listing-struct"></a>
```
struct Listing {
    address signer;
    address contractAddress;
    uint256 tokenId;
    uint256 expiration;
    uint256[3] nonces;
    uint256[] pricePerDay;
    uint256[] maxDays;
    uint256[] minDays;
    address target;
    bytes signature;
}
```

- signer: The address of the signer, in this case, my address.
- contractAddress: The address of the LAND contract.
- tokenId: The token id of the LAND I want to rent.
- expiration: The timestamp from which the signature will not be valid anymore.
- nonces: The three of them have to match the current nonces in the Rentals contract to be valid. Find more about it in the [Nonces](#nonces) section.
- pricePerDay: Defines the price per day for the min and max days range in the same index.
- maxDays: max amount of days the LAND can be rented for a given index.

> - minDays: min amount of days the LAND can be rented for a given index.

Reviewer: this is a has?
```suggestion
- minDays: min amount of days the LAND has to be be rented for a given index.
```

meaning, can it be rented for less than `minDays`?

Author: No it can't

> - target: If defined, only the target address can use the signature, if not, anyone can do.

Reviewer: ```suggestion
- target: If defined, only the target address can use the signature, if not, anyone can.
```

Once both the signature and the listings conditions are ready, I, or whichever off-chain system handling this, can store them to make them 
available for interested users that want to rent my LAND.

The interested user can interact with the Rentals contract via `acceptListing` with this data to initialize the rent. You can see more about [Accept Listing](#accepting-a-listing) in its corresponding section.

## Accepting a Listing

In the case that I am interested in renting a LAND as a tenant I can do the following:

1 - Authorize the Rentals contract to transfer my MANA

Rentals are paid upfront once they are initialized. For this, the Rentals contract needs authorization to move my MANA to the owner of the LAND I'm renting and a small cut to the DAO.

To do this, I can simply call the `approve` function in the MANA contract with the address of the Rentals contract as the spender and a value of MANA greater or equal to the price I have to pay for the rental. I could set an incredibly large number as the value so with just 1 authorization, I can forget about doing it again in the future and just go to step 2 directly.

2 - Obtain a listing and its corresponding signature

> In order to accept a listing, I would need to get the listing conditions as well as a signature created by the onwer of the LAND I want to rent. The way I do this depends on the off-chain system the owner has used to store this information. But once I obtain them, I'm good to call the `acceptListing` function and start the rental.

Reviewer: ```suggestion
In order to accept a listing, I would need to get the listing conditions as well as a signature created by the owner of the LAND I want to rent. This depends on the off-chain system the owner has used to store the information. But once I obtain them, I'm good to call the `acceptListing` function and start the rental.
```

> 2 - Call `acceptListing` in the Rentals contract

Reviewer: ```suggestion
3 - Call `acceptListing` in the Rentals contract
```

Once the previous information is at hand, I can finally call the function to accept the listing.

```
function acceptListing(
    Listing calldata _listing,
    address _operator,
    uint256 _index,
    uint256 _rentalDays,
    bytes32 _fingerprint
) external
```

- listing: [Listing](#listing-struct) data containing both the conditions and the signature of owner of the LAND.
- operator: The address that will be given operator permitions via the `setUpdateOperator` in the LAND contract. In other words, the user that I want being able to deploy scenes in that LAND, it could be myself if I wanted to but maybe I want someone else to do it.
- index: Remember that the Listing contains arrays for pricePerDay, maxDays and minDays? This value determines which index of those arrays I want to choose for this rental.
- rentalDays: The amount of days I want to rent this LAND. It has to be in range between the min and max days of the index I choose or else it fails.
- fingerprint: In the case of an Estate, this indicates the fingerprint it should have when renting it to me. This prevents the owner to front run me and remove LAND from it before renting it.

> If everything is correct, MANA equivalent to the pricePerDay index I've selected, times the rentalDays I provided will be transfered from my address to the lessor (and a fee to the DAO) and the provided operator will start being able to deploy scenes to that LAND.

Reviewer 2: fee to the FeeCollector, the `DAO` is an agent that may change.

## Offers

> This is the opposite of a [Listing](#listing). In this case, me, as a user interested in renting a certain LAND, would like to sign an Offer for said LAND so the owner has the possibility of renting it to me.

Reviewer 2: LAND or Estate

> This is the contrary of a [Listing](#listing). In this case, me, as a user interested of renting a certain LAND, would like to sign an Offer for said LAND so the owner has the possibility of renting it to me.

Reviewer: ```suggestion
This is the opposite of a [Listing](#listing). In this case, me, as a user interested in renting a certain LAND, would like to sign an Offer for said LAND so the owner has the possibility of renting it to me.
```

    address operator;
    bytes32 fingerprint;
    bytes signature;
}
```

- signer: The address of the signer, in this case, my address.
- contractAddress: The address of the LAND contract.
- tokenId: The token id of the LAND I want to rent.
- expiration: The timestamp from which the signature will not be valid anymore.
- nonces: The three of them have to match the current nonces in the Rentals contract to be valid. Find more about it in the [Nonces](#nonces) section.
- pricePerDay: Defines the price per day I'm willing to pay for renting the LAND
- rentalDays: The amount of days I want to rent the LAND
- operator: The address that will be given operator permissions to deploy scenes to that LAND
- fingerprint: If the asset would be an Estate, the fingerprint of that Estate should match the one I'm providing so the owner cannot remove LAND from it before renting it to me.
- signature: The signature created by signing all the previous data

There are various ways of signing these conditions. One of them can be achieved by using ethers as seen in the [./test/utlils/rentals.ts](https://github.com/decentraland/rentals-contract/blob/main/test/utils/rentals.ts#L81) utility file.

> Once both the signature and the listings conditions are ready, I, or whichever off-chain system handling this, can store them to make them 

Reviewer 2: the order conditions?

> Once both the signature and the listings conditions are ready, I, or whichever off-chain handling this, can store them to make them 

Reviewer: ```suggestion
Once both the signature and the listings conditions are ready, I, or whichever off-chain system handling this, can store them to make them 
```

not sure about this one, maybe missing `system`?

available for the interested owner of that LAND to accept it.

If the owner of the the LAND is interested in the offer, they can interact with the Rentals contract via `acceptOffer` or by sending the asset to the Rentals contract via the `safeTransferFrom` function in the LAND contract, providing the Offer in the last `bytes` parameter. More info on this can be found in the [Accepting an Offer](#accepting-an-offer) section.

## Accepting an Offer

> Me, as the current owner of a LAND, see that there is an offer by a much generous user that wants to rent my LAND for a year and for a lot of MANA.

Reviewer: This is a bit weird as `Me` does not function the same way as `I` in this context. The clarification between the comas make it seem as if the verb should be in third person (sees) but the subject (me) is in first. Thing is that `Me see` is not correct english.

maybe 
```suggestion
When I, as the current owner of a LAND, see that there is an offer by a much generous user that wants to rent my LAND for a year and for a lot of MANA.
```

> There are 2 ways in which, with the Offer condition and the offer signature in my possesion, can initialize the rent.

Reviewer: ```suggestion
There are 2 ways in which I, with the Offer condition and the offer signature in my possesion, can initialize the rent.
```

1 - Calling the `acceptListing` function in the Rentals contract. 

This function, similarly to Listing, requires that I, as the owner, [authorize](#authorize-rentals-contract) the Rentals contract to transfer the LAND.

Once that is out of the way, I can call `acceptListing` function to initialize the rent.

This function only receives one parameter as you can see in the following line:

```
function acceptOffer(Offer calldata _offer) external
```

- offer: [Offer](#offer-struct) containing the conditions and the signature of the offer.

> 2 - Sending the LAND to the Rentals cotract via LAND's `safeTransferFrom` function

Reviewer 2: `cotract` typo. => `Contract`

This is because in order to change or remove the update operator from a LAND, the `setUpdateOperator` function of the asset has to be manually called.

As the original owner of the asset, once the rental is over and I really don't want the operator defined by the tenant to keep deploying stuff after the rental period ends, I could do the following:

- Set the update operator manually to one of my liking as instructed [here](#changing-the-update-operator-of-the-asset).
- Claim the asset back as instructed [here](#claiming-back-the-asset).
- Accept a new Offer as instructed [here](#accepting-an-offer), changing the operator of the LAND to the new tenant's operator.
- Having a Listing accepted by a new tenant as instructed [here](#accepting-a-listing), changing the operator of the LAND to the new tenant's operator.

Another important thing to keep in mind is that there is NO need to claim the asset back once the rental is over to rent it again. 

After the rent of a LAND is over, the lessor can accept and offer or have a listing accepted and the LAND will be re-rented. This is useful because it saves the original owner from some extra transactions.

## Extending a rental

> BEFORE a rental is over, a new listing or offer for the LAND can be accepted as long as the lessor and tenant in said offer/listing are the same.

Reviewer 2: Can we explain how to do that by using the target parameter in the Listing struct?

Doing so will extend the end date of the rental by the amount of rental days defined.

## Nonces

> The Rentals contract contains various nonces of different types used to verify if a signature is valid. They can be modified used to [Invalidate Signatures](#invalidating-signatures) in many levels.

Reviewer: This phrase is a bit weird to me:
`They can be modified used to [Invalidate Signatures](#invalidating-signatures) in many levels.`

I don't really get what's trying to say

Author: Modified should not be there

- contractNonce: A nonce that once its changed by the owner of the contract, will make all signatures signed with the previous value invalid.

> - signerNonce: A nonce that once its changed by sender of the transaction, will make all signatures signed by the sender invalid.

Reviewer: ```suggestion
- signerNonce: A nonce that once its changed by the sender of the transaction, will make all signatures signed by the sender invalid.
```

> - assetNonce: A nonce that once its changed by sender of the transaction, will make all signatures signed by the sender for a given asset invalid.

Reviewer: ```suggestion
- assetNonce: A nonce that once its changed by the sender of the transaction, will make all signatures signed by the sender for a given asset invalid.
```

> The `uint256[3] nonces;` field in the [Listing](#listing-struct) struct, in order to be valid. must be conformed by the current [contactNonce, signerNonce, assetNonce].

Reviewer: ```suggestion
The `uint256[3] nonces;` field in the [Listing](#listing-struct) struct, in order to be valid, must be conformed by the current [contactNonce, signerNonce, assetNonce].
```

For example, if I wanted to sign a listing, I would need to ask the Rentals contract for these values via the public variables exposed:

```
uint256 public contractNonce;
mapping(address => uint256) public signerNonce;
mapping(address => mapping(uint256 => mapping(address => uint256))) public assetNonce;
```

In pseudo js code, if I wanted to rent a LAND, I would need to obtain the nonces doing something like this:

> ```

Reviewer: Maybe we can add `js` or `ts` here?

```
```js
or
```ts
```

Result:

```ts
const contractNonce = await rentalsContract.contractNonce();
const signerNonce = await rentalsContract.signerNonce(myAddress);
const assetNonce = await rentalsContract.contractNonce(landContractAddress, landTokenId, myAddress);
const listing = {
  ...otherParams,
  nonces: [contractNonce, signerNonce, assetNonce]
}
```

}
```

## Invalidating Signatures

Users can invalidate signatures in different ways by updating the contractNonce, the signerNonce and the assetNonce.

They can do so by calling the following functions in the Rentals contract.

```
function bumpContractNonce() external onlyOwner
function bumpSignerNonce() external
function bumpAssetNonce(address _contractAddress, uint256 _tokenId) external
```

`bumpContractNonce` can only be called by the owner of the Rentals contract, and will be used only on an emergency to invalidate all signatures created with the current nonce.

`bumpSignerNonce` can be called by any user, and will invalidate all signatures created by that user with the current nonce.

`bumpAssetNonce` can be called by any user, and will invalidate all signatures created by that user for a given asset with the current nonce.

The asset nonce is always bumped for both the lessor and the tenant once a rent is initialized to prevent any other listings or offers with the current nonce to be usable. This is a safety meassure to ensure that the least amount of usable signatures are available off-chain in the case of a signature storage breach.

> In the case of the signer nonce, imagine that I signed a lot of listings or offers with the current nonce but for some reason I don't trust the way they are stored, I could just call the asset nonce bump to invalidate them all at once.

Reviewer 2: What is the case of the contract nonce? Should we need to add it too?
